hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestLogRoller.java

```diff
@@ -43,53 +49,104 @@
+    ROLLER = new LogRoller(services);
+    ROLLER.start();
   }
 
   @After
   public void tearDown() throws Exception {
+    ROLLER.close();
+    FS.close();
     TEST_UTIL.shutdownMiniCluster();
   }
 
   @Test
   public void testRemoveClosedWAL() throws Exception {
-    HRegionServer rs = TEST_UTIL.getMiniHBaseCluster().getRegionServer(0);
-    Configuration conf = rs.getConfiguration();
-    LogRoller logRoller = TEST_UTIL.getMiniHBaseCluster().getRegionServer(0).getWalRoller();
-    int originalSize = logRoller.getWalNeedsRoll().size();
-    FSHLog wal1 = new FSHLog(rs.getWALFileSystem(), rs.getWALRootDir(),
-        AbstractFSWALProvider.getWALDirectoryName(rs.getServerName().getServerName()), conf);
-    logRoller.addWAL(wal1);
-    FSHLog wal2 = new FSHLog(rs.getWALFileSystem(), rs.getWALRootDir(),
-      AbstractFSWALProvider.getWALDirectoryName(rs.getServerName().getServerName()), conf);
-    logRoller.addWAL(wal2);
-    FSHLog wal3 = new FSHLog(rs.getWALFileSystem(), rs.getWALRootDir(),
-      AbstractFSWALProvider.getWALDirectoryName(rs.getServerName().getServerName()), conf);
-    logRoller.addWAL(wal3);
-
-    assertEquals(originalSize + 3, logRoller.getWalNeedsRoll().size());
-    assertTrue(logRoller.getWalNeedsRoll().containsKey(wal1));
-
-    wal1.close();
-    Thread.sleep(2 * logRollPeriod);
-
-    assertEquals(originalSize + 2, logRoller.getWalNeedsRoll().size());
-    assertFalse(logRoller.getWalNeedsRoll().containsKey(wal1));
-
-    wal2.close();
-    wal3.close();
-    Thread.sleep(2 * logRollPeriod);
-
-    assertEquals(originalSize, logRoller.getWalNeedsRoll().size());
+    assertEquals(0, ROLLER.getWalNeedsRoll().size());
+    for (int i = 1; i <= 3; i++) {
+      FSHLog wal = new FSHLog(FS, ROOT_DIR, LOG_DIR, ARCHIVE_DIR, CONF, null,
+        true, WAL_PREFIX, getWALSuffix(i));
+      ROLLER.addWAL(wal);
+    }
+
+    assertEquals(3, ROLLER.getWalNeedsRoll().size());
+    Iterator<WAL> it = ROLLER.getWalNeedsRoll().keySet().iterator();
+    WAL wal = it.next();
+    assertTrue(ROLLER.getWalNeedsRoll().containsKey(wal));
+
+    wal.close();
+    Thread.sleep(LOG_ROLL_PERIOD + 5000);
+
+    assertEquals(2, ROLLER.getWalNeedsRoll().size());
+    assertFalse(ROLLER.getWalNeedsRoll().containsKey(wal));
+
+    wal = it.next();
+    wal.close();
+    wal = it.next();
+    wal.close();
+    Thread.sleep(LOG_ROLL_PERIOD + 5000);
+
+    assertEquals(0, ROLLER.getWalNeedsRoll().size());
+  }
+
+  /**
+   * verify that each wal roll separately
+   */
+  @Test
+  public void testRequestRollWithMultiWal() throws Exception {
+    // add multiple wal
+    Map<FSHLog, Path> wals = new HashMap<>();
+    for (int i = 1; i <= 3; i++) {
+      FSHLog wal = new FSHLog(FS, ROOT_DIR, LOG_DIR, ARCHIVE_DIR, CONF, null,
+        true, WAL_PREFIX, getWALSuffix(i));
+      wal.init();
+      wals.put(wal, wal.getCurrentFileName());
+      ROLLER.addWAL(wal);
+      Thread.sleep(3000);
```

Comment:
We need a 3 sec sleep here? !

Reply:
We want 3 WALs will different periodic roll time period?  Why we need? If so, can we limit the time to be 1 sec or lesser?